src/box/rt_lua_call_access.c

```diff
@@ -0,0 +1,101 @@
+#include "rt_lua_call_access.h"
+#include "diag.h"
+#include "assoc.h"
+#include "user.h"
+#include "session.h"
+#include "../lua/init.h" /* tarantool_lua_is_builtin_global */
+
+static struct mh_access_t *user_rt_access_cache = NULL;
+static struct mh_strnptr_t *user_rt_universe_access_cache = NULL;
+
+const char *
+user_name_by_id(uint32_t uid, uint32_t *name_len)
+{
+	struct user *user = user_find(uid);
+	const char *name = user->def->name;
+	*name_len = strlen(name);
+	return name;
+}
+
+void
+rt_lua_call_access_init(void)
+{
+	user_rt_access_cache = mh_access_new();
+	user_rt_universe_access_cache = mh_strnptr_new();
+}
+
+void
+rt_lua_call_access_destroy(void)
+{
+	mh_access_delete(user_rt_access_cache);
+	mh_strnptr_delete(user_rt_universe_access_cache);
+}
+
+void
+grant_rt_access(const char *uname, uint32_t uname_len,
+		const char *fname, uint32_t fname_len)
+{
+	/* Grant universe access. */
+	if (is_universe(fname, fname_len)) {
+		uint32_t hash = mh_strn_hash(uname, uname_len);
+		const struct mh_strnptr_node_t uid_node =
+			{ uname, uname_len, hash, NULL };
+		struct mh_strnptr_node_t repl_uid_node = { NULL, 0, 0, NULL };
+		struct mh_strnptr_node_t *prepl_uid_node = &repl_uid_node;
+		mh_strnptr_put(user_rt_universe_access_cache, &uid_node,
+			       &prepl_uid_node, NULL);
+		return;
+	}
+
+	/* Grant access to lua_call function. */
+	uint32_t hash = mh_access_hash(uname, uname_len, fname, fname_len);
+	const struct mh_access_node_t access_node =
+		{ uname, uname_len, fname, fname_len, hash, RT_ACCESS };
```

Comment:
It may be convenient to have a `mh_strn_strn_node_create()` function that calculates the hash:

```c
struct mh_strn_strn_node_t access_node;
mh_strn_strn_node_create(&access_node, uname, uname_len, fname, fname_len);
```

Reply:
I see no problem (if we do it after all the needed declarations):

```diff
diff --git a/src/box/rt_access.c b/src/box/rt_access.c
index acfd29a2c4..a1417039d6 100644
--- a/src/box/rt_access.c
+++ b/src/box/rt_access.c
@@ -5,19 +5,6 @@
 static struct mh_strn_strn_t *user_rt_access = NULL;
 static struct mh_strnptr_t *user_rt_universe_access = NULL;
 
-void
-mh_strn_strn_node_create(struct mh_strn_strn_node_t *node,
-			 const char *s1, uint32_t s1_len,
-			 const char *s2, uint32_t s2_len)
-{
-	uint32_t hash = mh_strn_strn_hash(s1, s1_len, s2, s2_len);
-	node->s1 = s1;
-	node->s1_len = s1_len;
-	node->s2 = s2;
-	node->s2_len = s2_len;
-	node->hash = hash;
-}
-
 void
 rt_access_init(void)
 {
diff --git a/src/lib/core/assoc.h b/src/lib/core/assoc.h
index 1e91b19794..bc53d360ed 100644
--- a/src/lib/core/assoc.h
+++ b/src/lib/core/assoc.h
@@ -231,6 +231,19 @@ struct mh_strn_strn_node_t {
 #define mh_cmp_key(a, b, arg) mh_cmp(a, b, arg)
 #include "salad/mhash.h"
 
+static inline void
+mh_strn_strn_node_create(struct mh_strn_strn_node_t *node,
+			 const char *s1, uint32_t s1_len,
+			 const char *s2, uint32_t s2_len)
+{
+	uint32_t hash = mh_strn_strn_hash(s1, s1_len, s2, s2_len);
+	node->s1 = s1;
+	node->s1_len = s1_len;
+	node->s2 = s2;
+	node->s2_len = s2_len;
+	node->hash = hash;
+}
+
 #if defined(__cplusplus)
 } /* extern "C" */
 #endif /* defined(__cplusplus) */
```

(We can also move `mh_strn_strn_hash` below to place all the `*_strn_strn_*` functions together.)

Reply:
I removed `mh_strn_strn_node_create` because it is now used in one place and I don't think it is necessary.